pytorch_lightning/overrides/data_parallel.py

```diff
@@ -63,7 +65,39 @@ def forward(self, *inputs, **kwargs):
 
         replicas = self.replicate(self.module, self.device_ids[:len(inputs)])
         outputs = self.parallel_apply(replicas, inputs, kwargs)
-        return self.gather(outputs, self.output_device)
+        return self.gather(outputs)
+
+    def gather(self, outputs):
+        r"""
+        Override the gather method to support scalars as well.
+        """
+        def gather_map(outputs):
+            out = outputs[0]
+            if isinstance(out, torch.Tensor):
+                return Gather.apply(self.output_device, self.dim, *outputs)
+
+            elif out is None:
+                return None
+
+            elif isinstance(out, dict):
+                if not all((len(out) == len(d) for d in outputs)):
+                    raise ValueError('All dicts must have the same number of keys')
+                return type(out)(((k, gather_map([d[k] for d in outputs]))
+                                  for k in out))
+
+            elif isinstance(out, Iterable):
+                return type(out)(map(gather_map, zip(*outputs)))
+
+            # assume "out" is a scalar
+            return outputs
+
+        # Recursive function calls like this create reference cycles.
+        # Setting the function to None clears the refcycle.
+        try:
+            res = gather_map(outputs)
+        finally:
```

Comment:
which exception is raised here specifically?

Reply:
It could be any exception. For example, look at the issues that I mentioned in the description. This part of code is from the original PyTorch implementation. I think the reason they used a try-finally block here is to break the reference cycle, even when an exception is raised or caught from outside.

Reply:
alright, in this case I think this is good.